pkg/core/gitea/client.go

```diff
@@ -0,0 +1,309 @@
+			break
+		}
+	}
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to create application")
+	}
+
+	return nil
+}
+
+// create git repository with given name under given org
+func (gac *GiteaAdminClient) CreateRepo(c *codeset.Codeset) error {
+	_, resp, err := gac.giteaClient.GetRepo(c.Project, c.Name)
+	if resp == nil && err != nil {
+		return errors.Wrap(err, "Failed to make get repo request")
+	}
+
+	if resp.StatusCode == 200 {
+		log.Printf("Repository '%s' already exists under '%s'", c.Name, c.Project)
+		return nil
+	}
+
+	log.Printf("Creating repository '%s' under '%s'...", c.Name, c.Project)
+	_, _, err = gac.giteaClient.CreateOrgRepo(c.Project, gitea.CreateRepoOption{
+		Name:          c.Name,
+		AutoInit:      true,
+		Private:       true,
+		DefaultBranch: "main",
+		Description:   *c.Description,
+	})
+
+	if err != nil {
+		return errors.Wrap(err, "Failed to create repository")
+	}
+
+	return nil
+}
+
+// Add topics to given repository
+func (gac *GiteaAdminClient) AddRepoTopics(org, name string, labels []string) error {
+	for _, label := range labels {
+		_, err := gac.giteaClient.AddRepoTopic(org, name, label)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Create webhook for given repository and wire it to tekton listener
+func (gac *GiteaAdminClient) CreateRepoWebhook(org, name string) error {
+	hooks, _, err := gac.giteaClient.ListRepoHooks(org, name, gitea.ListHooksOptions{})
+	if err != nil {
+		return errors.Wrap(err, "Failed to list webhooks")
+	}
+
+	for _, hook := range hooks {
+		url := hook.Config["url"]
+		if url == config.StagingEventListenerURL {
+			log.Printf("Webhook for '%s' already exists", name)
+			return nil
+		}
+	}
+
+	log.Printf("Creating Webhook for '%s' under '%s'...", name, org)
+	gac.giteaClient.CreateRepoHook(org, name, gitea.CreateHookOption{
+		Active:       true,
+		BranchFilter: "*",
+		Config: map[string]string{
+			"secret":       config.HookSecret,
+			"http_method":  "POST",
+			"url":          config.StagingEventListenerURL,
+			"content_type": "json",
+		},
+		Type: "gitea",
+	})
+
+	return nil
+}
+
+// Prepare the org, repository, and create user that clients can use for pushing
+func (gac *GiteaAdminClient) PrepareRepo(code *codeset.Codeset) error {
+
+	err := gac.CreateOrg(code.Project)
+	if err != nil {
+		return errors.Wrap(err, "Create org failed")
+	}
+
+	err = gac.CreateUser(code.Project)
+	if err != nil {
+		return errors.Wrap(err, "Create FuseML user failed")
+	}
+
+	err = gac.CreateRepo(code)
+	if err != nil {
+		return errors.Wrap(err, "Create repo failed")
+	}
+
+	err = gac.AddRepoTopics(code.Project, code.Name, code.Labels)
+	if err != nil {
+		return errors.Wrap(err, "Failed to add topics to repository")
+	}
+
+	err = gac.CreateRepoWebhook(code.Project, code.Name)
+	if err != nil {
+		return errors.Wrap(err, "Creating webhook failed")
+	}
+	return nil
+}
+
+// simple check if a string is present in a slice
+func contains(s []string, str string) bool {
```

Comment:
One more reason to use a dictionary/map to represent labels :)

Reply:
Well, maybe. The problem here is not with _our_ structure, but with the gitea API that returns `[]string` .

(Than you might want to say, let's not use git for labels - and I guess we'll see if we need some extra store...)